pennylane/interfaces/torch.py

```diff
@@ -16,13 +16,48 @@
 compatible quantum nodes.
 """
 # pylint: disable=redefined-outer-name,arguments-differ
+from collections import Iterable
 import inspect
 from functools import partial
+import numbers
 
 import numpy as np
 import torch
+from torch.autograd.function import once_differentiable
 
-from pennylane.utils import unflatten
+from pennylane.utils import _flatten
+
+
+def unflatten_torch(flat, model):  # pragma: no cover
+    """Restores an arbitrary nested structure to a flattened Torch tensor.
+
+    Args:
+        flat (torch.Tensor): 1D tensor of items
+        model (array, Iterable, Number): model nested structure
+
+    Returns:
+        Union[torch.Tensor, list], array: first elements of flat arranged into the nested
+        structure of model, unused elements of flat
+
+    Raises:
+        TypeError: if ``model`` contains an object of unsupported type
+    """
+    if isinstance(model, (numbers.Number, str)):
+        return flat[0], flat[1:]
+
+    if isinstance(model, torch.Tensor):
+        idx = model.numel()
+        res = torch.view(flat[:idx], model.shape)
+        return res, flat[idx:]
+
+    if isinstance(model, Iterable):
+        res = []
+        for x in model:
+            val, flat = unflatten_torch(flat, x)
+            res.append(val)
+        return res, flat
```

Comment:
I left the same comment on #626 😆 

In `pennylane.utils`, we have an explicit validation for the general-purpose unflatten function:
```python
def unflatten(flat, model):
    res, tail = _unflatten(np.asarray(flat), model)
    if len(tail) != 0:
        raise ValueError("Flattened iterable has more elements than the model.")
    return res
```

Here, however, there is a slight usage subtlety: `unflatten_torch()` is only used in the Torch interface, and only for backpropagation. As a result:

* Since Torch performs the backpropagation, there is no combination of inputs that would ever cause `len(flat) != 0`, and if it _did_, would likely be emblematic of a bug in PyTorch. This makes it impossible to test. Further, instead of input validation, this is more of a 'should never happen' assertion.

* Including an assertion within `TorchQNode.backward()` (e.g., `assert len(flat) != 0, "Something went wrong!"`) causes issues with the PyTorch functional interface, and is still very difficult to test, since the backward call is 'invisible' to pytest and pytest-cov.

@trbromley had a chat about this, and decided to remove the validation/assertion.

Reply:
wait did I misinterpret your question

Reply:
> wait did I misinterpret your question

Maybe? :man_shrugging: 

The thing I was pointing out is that this torch function has an extra `for` loop that the pennylane `unflatten` does not have. So the "leftover" `flat` is only really the leftover of the last item in `model` (earlier items might have "tails", but they are not accounted for anywhere